src/device-manager/WeaveDataManagementClient.cpp

```diff
@@ -966,35 +1142,63 @@ void WdmClient::ClientEventCallback(void * const aAppState, SubscriptionClient::
             else
             {
                 traitListLen = pWdmClient->mSinkCatalog.Size();
-                VerifyOrExit(traitListLen != 0, err = WEAVE_ERROR_INVALID_LIST_LENGTH);
             }
             WeaveLogDetail(DataManagement, "prepare to subscribe %d trait data sink", traitListLen);
 
-            if (pWdmClient->mpPublisherPathList == NULL)
-            {
-                delete[] pWdmClient->mpPublisherPathList;
-            }
-
-            pWdmClient->mpPublisherPathList = new TraitPath[traitListLen];
-
-            if (needSubscribeSpecificTrait)
+            if (!pWdmClient->mEventFetching)
             {
-                pathListLen = 1;
-                err = pWdmClient->mSinkCatalog.PrepareSubscriptionSpecificPathList(pWdmClient->mpPublisherPathList, traitListLen,
-                                                                                   handle);
+                if (pWdmClient->mpPublisherPathList != NULL)
+                {
+                    delete[] pWdmClient->mpPublisherPathList;
+                    pWdmClient->mpPublisherPathList = NULL;
+                }
+
+                if (traitListLen > 0)
+                {
+                    pWdmClient->mpPublisherPathList = new TraitPath[traitListLen];
+                }
+
+                if (needSubscribeSpecificTrait)
+                {
+                    pathListLen = 1;
+                    err = pWdmClient->mSinkCatalog.PrepareSubscriptionSpecificPathList(pWdmClient->mpPublisherPathList, traitListLen,
+                                                                                    handle);
+                }
+                else
+                {
+                    err = pWdmClient->mSinkCatalog.PrepareSubscriptionPathList(pWdmClient->mpPublisherPathList, traitListLen,
+                                                                            pathListLen);
+                }
+                SuccessOrExit(err);
             }
             else
             {
-                err = pWdmClient->mSinkCatalog.PrepareSubscriptionPathList(pWdmClient->mpPublisherPathList, traitListLen,
-                                                                           pathListLen);
+                // When mEventFetching is true, we will not subscribe to any traits
```

Comment:
Please put this logic in one function and reduce the current function size a little bit.